pkg/clustermesh/mcsapi/serviceimport_controller.go

```diff
@@ -0,0 +1,489 @@
+				appProtocol = *port.AppProtocol
+			}
+			key := portMergeKey{
+				Protocol:    port.Protocol,
+				AppProtocol: appProtocol,
+				Port:        port.Port,
+			}
+			val, ok := globalPorts[key]
+			if !ok || mcsAPISvcSpec.ExportCreationTimestamp.Before(&val.ExportCreationTimestamp) {
+				globalPorts[key] = portMergeValue{
+					ServicePort:             port,
+					clusterName:             cluster,
+					ExportCreationTimestamp: mcsAPISvcSpec.ExportCreationTimestamp,
+				}
+			}
+		}
+	}
+	return globalPorts
+}
+
+func mergedPortsToMCSPorts(mergedPorts mergedPorts) []mcsapiv1alpha1.ServicePort {
+	ports := make([]mcsapiv1alpha1.ServicePort, 0, len(mergedPorts))
+	for _, port := range mergedPorts {
+		ports = append(ports, port.ServicePort)
+	}
+	return ports
+}
+
+func getServiceImportStatus(svcExportByCluster operator.ServiceExportsByCluster) mcsapiv1alpha1.ServiceImportStatus {
+	clusters := []mcsapiv1alpha1.ClusterStatus{}
+	clusterNames := maps.Keys(svcExportByCluster)
+	sort.Strings(clusterNames)
+	for _, clusterName := range clusterNames {
+		clusters = append(clusters, mcsapiv1alpha1.ClusterStatus{
+			Cluster: clusterName,
+		})
+	}
+	return mcsapiv1alpha1.ServiceImportStatus{Clusters: clusters}
+}
+
+func checkConflictExport(svcExportByCluster operator.ServiceExportsByCluster, clusterName, oldestClusterName string) string {
+	localSvcSpec := svcExportByCluster[clusterName]
+	oldestSvcSpec := svcExportByCluster[oldestClusterName]
+	clusterCount := len(svcExportByCluster)
+
+	getTypeFunc := func(svcSpec *mcsapitypes.MCSAPIServiceSpec) string {
+		return string(svcSpec.Type)
+	}
+	getSessionAffinityFunc := func(svcSpec *mcsapitypes.MCSAPIServiceSpec) string {
+		return string(svcSpec.SessionAffinity)
+	}
+	getSessionAffinityConfigFunc := func(svcSpec *mcsapitypes.MCSAPIServiceSpec) string {
+		if svcSpec.SessionAffinityConfig == nil ||
+			svcSpec.SessionAffinityConfig.ClientIP == nil ||
+			svcSpec.SessionAffinityConfig.ClientIP.TimeoutSeconds == nil {
+			return ""
+		}
+		return string(*svcSpec.SessionAffinityConfig.ClientIP.TimeoutSeconds)
+	}
+	conflictMap := map[string]func(svcSpec *mcsapitypes.MCSAPIServiceSpec) string{
+		"type":                           getTypeFunc,
+		"sessionAffinity":                getSessionAffinityFunc,
+		"sessionAffinityConfig.clientIP": getSessionAffinityConfigFunc,
+	}
+	for fieldName, getFieldAsStringFunc := range conflictMap {
+		localField := getFieldAsStringFunc(localSvcSpec)
+		oldestSvcField := getFieldAsStringFunc(oldestSvcSpec)
+		if localField == oldestSvcField {
+			continue
+		}
+
+		conflictCount := 0
+		for _, currentSvcSpec := range svcExportByCluster {
+			if oldestSvcField != getFieldAsStringFunc(currentSvcSpec) {
+				conflictCount += 1
+			}
+		}
+
+		return fmt.Sprintf(
+			"Conflicting %s. %d/%d clusters disagree. Using \"%s\" from oldest service export in \"%s\".",
+			fieldName, conflictCount, clusterCount, oldestSvcField, oldestClusterName,
+		)
```

Comment:
In theory yes that sounds like a good idea, in practice I am not sure how readable it would be. I haven't see yet some controllers that were concatenating multiple errors in a condition somehow (but I am not sure if it's because of the complexity it would bring or because user wise it is not readable :/)

Reply:
Yeah, I'm also a bit doubtful on how the concatenated messages would be rendered. In any case, I'm perfectly fine with the current approach.